cmd/harmony/config/config.go

```diff
@@ -255,67 +255,102 @@ func parseNetworkType(nt string) nodeconfig.NetworkType {
 	default:
 		return defaultElseSyncConfig
 	}
 }
 
-var dumpConfigCmd = &cobra.Command{
-	Use:   "dumpconfig [config_file]",
-	Short: "dump the config file for harmony binary configurations",
-	Long:  "dump the config file for harmony binary configurations",
-	Args:  cobra.MinimumNArgs(1),
-	Run: func(cmd *cobra.Command, args []string) {
-		nt := getNetworkType(cmd)
-		config := getDefaultHmyConfigCopy(nt)
-
-		if err := writeHarmonyConfigToFile(config, args[0]); err != nil {
-			fmt.Println(err)
-			os.Exit(128)
-		}
-	},
-}
-
-func registerDumpConfigFlags() error {
-	return cli.RegisterFlags(dumpConfigCmd, []cli.Flag{networkTypeFlag})
+func GetDefaultDNSSyncConfig(nt nodeconfig.NetworkType) DNSSyncConfig {
+	config := DNSSyncConfig{
+		LegacySyncing: false,
+		DNSZone:       nodeconfig.GetDefaultDNSZone(nt),
+		DNSPort:       nodeconfig.GetDefaultDNSPort(nt),
+		LegacyServer:  true,
+		LegacyClient:  true,
+	}
+	switch nt {
+	case nodeconfig.Mainnet:
+	case nodeconfig.Testnet:
+	case nodeconfig.Localnet:
+	default:
+		config.LegacyClient = false
+	}
+	return config
 }
 
-func loadHarmonyConfig(file string) (harmonyConfig, error) {
+func LoadHarmonyConfig(file string) (HarmonyConfig, error) {
 	b, err := ioutil.ReadFile(file)
 	if err != nil {
-		return harmonyConfig{}, err
+		return HarmonyConfig{}, err
 	}
-
-	var config harmonyConfig
-	if err := toml.Unmarshal(b, &config); err != nil {
-		return harmonyConfig{}, err
+	var header struct {
+		Version string
+	}
+	if err := toml.Unmarshal(b, &header); err != nil {
+		return HarmonyConfig{}, fmt.Errorf("reading config version: %w", err)
+	}
+	var config HarmonyConfig
+	// lexicographicaly compare
+	if header.Version <= v1TOMLConfigVersion {
+		config, err = loadHarmonyConfigV1(b)
+	} else {
+		err = toml.Unmarshal(b, &config)
+	}
+	if err != nil {
+		return HarmonyConfig{}, err
 	}
 
 	// Correct for old config version load (port 0 is invalid anyways)
 	if config.HTTP.RosettaPort == 0 {
-		config.HTTP.RosettaPort = defaultConfig.HTTP.RosettaPort
+		config.HTTP.RosettaPort = DefaultConfig.HTTP.RosettaPort
 	}
 	if config.P2P.IP == "" {
-		config.P2P.IP = defaultConfig.P2P.IP
+		config.P2P.IP = DefaultConfig.P2P.IP
 	}
 	if config.Prometheus == nil {
-		config.Prometheus = defaultConfig.Prometheus
+		config.Prometheus = DefaultConfig.Prometheus
 	}
 	return config, nil
 }
 
-func writeHarmonyConfigToFile(config harmonyConfig, file string) error {
+func WriteHarmonyConfigToFile(config HarmonyConfig, file string) error {
 	b, err := toml.Marshal(config)
 	if err != nil {
 		return err
 	}
 	return ioutil.WriteFile(file, b, 0644)
 }
+
+// BackupAndUpgradeConfigToTheLatestVersion backups previous version of config file and persists on the disk the latest format.
+// It returns file name of the backup.
+func BackupAndUpgradeConfigToTheLatestVersion(cfg HarmonyConfig, fileName string) (string, error) {
+	oldFileName := fileName + ".old"
+	for {
+		_, err := os.Stat(oldFileName)
```

Comment:
Potential forever loop and eats up all CPU resources. Would you please add a timely check logic and timeout here?

Reply:
Do you have any thoughts on this one?